### ISSUE TITLE
NUMA and irqbalance

### ISSUE THREAD

**Issue by nmenon** [Irqbalance/irqbalance#88]
On certain ARM SoCs which do not need NUMA, devicetree seems to describe the cputopology quite well. At the very first look, it seems that there is a constraint that irqbalance utility depends entirely on NUMA being enabled and fails to function on systems where SMP is present but NUMA is disabled.

Just curious on background on this.

**Comment by nhorman** [comment]
I'm not sure where you get that impression from.  Irqbalance will attempt to build a numa node tree when it querys the system, but if no numa nodes are found, all devices and irqs are assigned to the internally defined unspecified numa node, which treats all device<->cpu distances equally.  If thats not happening, its a bug, but its incorrect to say that irqbalance depends entirely on numa being enabled

**Comment by nsekhar** [comment]
Here are the logs with and without CONFIG_NUMA enabled.
irqbalance is that available in latest debian unstable for arm64 (1.5.0-3).
This is on ARM64 based TI AM65x.

With CONFIG_NUMA=y (v4.19 kernel)
```
root@charlie:~# cat /proc/interrupts 
           CPU0       CPU1       CPU2       CPU3       
  3:       3205       2408       2563       2239     GICv3  30 Level     arch_timer
  7:         11          0          0          0     GICv3 728 Level     42120000.i2c
  9:          0          0          0          0     GICv3 596 Level     40b00000.i2c
 10:       4291          0          0          0     GICv3  69 Level     32c00000.mailbox thr_011
 11:        402          0          0          0     GICv3 224 Level     2800000.serial
 15:          8          0          0          0     GICv3 232 Level     2000000.i2c
 16:          0          0          0          0     GICv3 233 Level     2010000.i2c
 17:          0          0          0          0     GICv3 234 Level     2020000.i2c
 18:          0          0          0          0     GICv3 235 Level     2030000.i2c
 20:        277          0          0          0     GICv3 168 Level     mmc0
 49:          0          0          0          0      GPIO  24 Edge    -davinci_gpio  GPIO Key USER1
 50:          0          0          0          0      GPIO  25 Edge    -davinci_gpio  0-0039
 52:          0          0          0          0      GPIO  27 Edge    -davinci_gpio  GPIO Key USER2
279:      15553          0       1299          0      INTA  16 Level     46000000.ethernet
280:      35730          0       2975          0      INTA  17 Level     46000000.ethernet
281:          0          0          0          0     GICv3 602 Level     46000000.ethernet
IPI0:      9453      17902      11917      11830       Rescheduling interrupts
IPI1:        54         74         98        118       Function call interrupts
IPI2:         0          0          0          0       CPU stop interrupts
IPI3:         0          0          0          0       CPU stop (for crash dump) interrupts
IPI4:         0          0          0          0       Timer broadcast interrupts
IPI5:         0          0          0          0       IRQ work interrupts
IPI6:         0          0          0          0       CPU wake-up interrupts
Err:          0
root@charlie:~# 

root@charlie:~# irqbalance --debug
Isolated CPUs: 00000000
Adaptive-ticks CPUs: 00000000
Banned CPUs: 00000000
Package 1:  numa_node 0 cpu mask is 0000000c (load 0)
        Cache domain 0:  numa_node is 0 cpu mask is 0000000c  (load 0) 
                CPU number 3  numa_node is 0 (load 0)
                CPU number 2  numa_node is 0 (load 0)
Package 0:  numa_node 0 cpu mask is 00000003 (load 0)
        Cache domain 1:  numa_node is 0 cpu mask is 00000003  (load 0) 
                CPU number 1  numa_node is 0 (load 0)
                CPU number 0  numa_node is 0 (load 0)
IRQ arch_timer(3) guessed as class 0
No directory /sys/devices/platform/42120000.i2c/: No such file or directory
IRQ 42120000.i2c(7) guessed as class 0
IRQ 42120000.i2c(7) guessed as class 0
IRQ 40b00000.i2c(9) guessed as class 0
IRQ thr_011(10) guessed as class 0
IRQ 2800000.serial(11) guessed as class 0
IRQ 2000000.i2c(15) guessed as class 0
IRQ 2010000.i2c(16) guessed as class 0
IRQ 2020000.i2c(17) guessed as class 0
IRQ 2030000.i2c(18) guessed as class 0
IRQ mmc0(20) guessed as class 0
IRQ USER1(49) guessed as class 0
IRQ 0-0039(50) guessed as class 0
IRQ USER2(52) guessed as class 0
IRQ 46000000.ethernet(279) guessed as class 5
No directory /sys/devices/platform/46000000.ethernet/: No such file or directory
IRQ 46000000.ethernet(279) guessed as class 0
IRQ 46000000.ethernet(279) guessed as class 0
IRQ 46000000.ethernet(280) guessed as class 5
No directory /sys/devices/platform/46000000.ethernet/: No such file or directory
IRQ 46000000.ethernet(280) guessed as class 0
IRQ 46000000.ethernet(280) guessed as class 0
IRQ 46000000.ethernet(281) guessed as class 5
No directory /sys/devices/platform/46000000.ethernet/: No such file or directory
IRQ 46000000.ethernet(281) guessed as class 0
IRQ 46000000.ethernet(281) guessed as class 0
Adding IRQ 3 to database
Adding IRQ 7 to database
Adding IRQ 9 to database
Adding IRQ 10 to database
Adding IRQ 11 to database
Adding IRQ 15 to database
Adding IRQ 16 to database
Adding IRQ 17 to database
Adding IRQ 18 to database
Adding IRQ 20 to database
Adding IRQ 49 to database
Adding IRQ 50 to database
Adding IRQ 52 to database
Adding IRQ 279 to database
Adding IRQ 280 to database
Adding IRQ 281 to database
NUMA NODE NUMBER: -1
LOCAL CPU MASK: ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,fffffffff

NUMA NODE NUMBER: 0
LOCAL CPU MASK: 0000000f




-----------------------------------------------------------------------------
Package 1:  numa_node 0 cpu mask is 0000000c (load 0)
        Cache domain 0:  numa_node is 0 cpu mask is 0000000c  (load 0) 
                CPU number 3  numa_node is 0 (load 0)
                CPU number 2  numa_node is 0 (load 0)
  Interrupt 281 node_num is -1 (other/0:0) 
  Interrupt 279 node_num is -1 (other/0:1) 
  Interrupt 50 node_num is -1 (other/0:0) 
  Interrupt 20 node_num is -1 (other/0:0) 
  Interrupt 17 node_num is -1 (other/0:0) 
  Interrupt 15 node_num is -1 (other/0:0) 
  Interrupt 10 node_num is -1 (other/0:0) 
  Interrupt 7 node_num is -1 (other/0:0) 
Package 0:  numa_node 0 cpu mask is 00000003 (load 0)
        Cache domain 1:  numa_node is 0 cpu mask is 00000003  (load 0) 
                CPU number 1  numa_node is 0 (load 0)
                CPU number 0  numa_node is 0 (load 0)
  Interrupt 280 node_num is -1 (other/0:85) 
  Interrupt 52 node_num is -1 (other/0:0) 
  Interrupt 49 node_num is -1 (other/0:0) 
  Interrupt 18 node_num is -1 (other/0:0) 
  Interrupt 16 node_num is -1 (other/0:0) 
  Interrupt 11 node_num is -1 (other/0:55) 
  Interrupt 9 node_num is -1 (other/0:0) 
  Interrupt 3 node_num is -1 (other/0:295) 



-----------------------------------------------------------------------------
Selecting irq 280 for rebalancing
Selecting irq 11 for rebalancing
Package 1:  numa_node 0 cpu mask is 0000000c (load 3214224)
        Cache domain 0:  numa_node is 0 cpu mask is 0000000c  (load 0) 
                CPU number 3  numa_node is 0 (load 0)
                CPU number 2  numa_node is 0 (load 0)
  Interrupt 281 node_num is -1 (other/1:0) 
  Interrupt 279 node_num is -1 (other/1:2) 
  Interrupt 50 node_num is -1 (other/1:0) 
  Interrupt 20 node_num is -1 (other/1:0) 
  Interrupt 17 node_num is -1 (other/1:0) 
  Interrupt 15 node_num is -1 (other/1:0) 
  Interrupt 10 node_num is -1 (other/1:0) 
  Interrupt 7 node_num is -1 (other/1:0) 
  Interrupt 280 node_num is -1 (other/2582368:94) 
  Interrupt 11 node_num is -1 (other/631856:23) 
Package 0:  numa_node 0 cpu mask is 00000003 (load 10000000)
        Cache domain 1:  numa_node is 0 cpu mask is 00000003  (load 10000000) 
                CPU number 1  numa_node is 0 (load 0)
                CPU number 0  numa_node is 0 (load 10000000)
  Interrupt 3 node_num is -1 (other/6785584:247) 
  Interrupt 52 node_num is -1 (other/1:0) 
  Interrupt 49 node_num is -1 (other/1:0) 
  Interrupt 18 node_num is -1 (other/1:0) 
  Interrupt 16 node_num is -1 (other/1:0) 
  Interrupt 9 node_num is -1 (other/1:0) 

# for i in $(seq 0 300)
> do
> grep . /proc/irq/$i/smp_affinity /dev/null 2>/dev/null
> done
/proc/irq/1/smp_affinity:f
/proc/irq/2/smp_affinity:f
/proc/irq/3/smp_affinity:f
/proc/irq/4/smp_affinity:f
/proc/irq/5/smp_affinity:f
/proc/irq/7/smp_affinity:c
/proc/irq/9/smp_affinity:3
/proc/irq/10/smp_affinity:c
/proc/irq/11/smp_affinity:c
/proc/irq/15/smp_affinity:c
/proc/irq/16/smp_affinity:3
/proc/irq/17/smp_affinity:c
/proc/irq/18/smp_affinity:3
/proc/irq/20/smp_affinity:c
/proc/irq/49/smp_affinity:f
/proc/irq/50/smp_affinity:f
/proc/irq/52/smp_affinity:f
/proc/irq/279/smp_affinity:c
/proc/irq/280/smp_affinity:c
/proc/irq/281/smp_affinity:c

root@charlie:~# lstopo-no-graphics
Machine (3990MB)
  Package L#0 + L2 L#0 (512KB)
    L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0 + PU L#0 (P#0)
    L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1 + PU L#1 (P#1)
  Package L#1 + L2 L#1 (512KB)
    L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2 + PU L#2 (P#2)
    L1d L#3 (32KB) + L1i L#3 (32KB) + Core L#3 + PU L#3 (P#3)
root@charlie:~# 
```
With CONFIG_NUMA=n (v4.19 kernel)

```
root@charlie:~# cat /proc/interrupts 
           CPU0       CPU1       CPU2       CPU3       
  3:       2880       2378       1770       1716     GICv3  30 Level     arch_timer
  7:         11          0          0          0     GICv3 728 Level     42120000.i2c
  9:          0          0          0          0     GICv3 596 Level     40b00000.i2c
 10:       4291          0          0          0     GICv3  69 Level     32c00000.mailbox thr_011
 11:        343          0          0          0     GICv3 224 Level     2800000.serial
 15:          8          0          0          0     GICv3 232 Level     2000000.i2c
 16:          0          0          0          0     GICv3 233 Level     2010000.i2c
 17:          0          0          0          0     GICv3 234 Level     2020000.i2c
 18:          0          0          0          0     GICv3 235 Level     2030000.i2c
 20:        241          0          0          0     GICv3 168 Level     mmc0
 49:          0          0          0          0      GPIO  24 Edge    -davinci_gpio  GPIO Key USER1
 50:          0          0          0          0      GPIO  25 Edge    -davinci_gpio  0-0039
 52:          0          0          0          0      GPIO  27 Edge    -davinci_gpio  GPIO Key USER2
279:      16687          0          0          0      INTA  16 Level     46000000.ethernet
280:      37825          0          0          0      INTA  17 Level     46000000.ethernet
281:          0          0          0          0     GICv3 602 Level     46000000.ethernet
IPI0:      7231      14937      14837      14437       Rescheduling interrupts
IPI1:        94         58        131        117       Function call interrupts
IPI2:         0          0          0          0       CPU stop interrupts
IPI3:         0          0          0          0       CPU stop (for crash dump) interrupts
IPI4:         0          0          0          0       Timer broadcast interrupts
IPI5:         0          0          0          0       IRQ work interrupts
IPI6:         0          0          0          0       CPU wake-up interrupts
Err:          0

# irqbalance --debug
This machine seems not NUMA capable.
Isolated CPUs: 00000000
Adaptive-ticks CPUs: 00000000
Banned CPUs: 00000000
Package 1:  numa_node cpu mask is 0000000c (load 0)
        Cache domain 0:  numa_node is cpu mask is 0000000c  (load 0) 
                CPU number 3  numa_node is (load 0)
                CPU number 2  numa_node is (load 0)
Package 0:  numa_node cpu mask is 00000003 (load 0)
        Cache domain 1:  numa_node is cpu mask is 00000003  (load 0) 
                CPU number 1  numa_node is (load 0)
                CPU number 0  numa_node is (load 0)
IRQ arch_timer(3) guessed as class 0
No directory /sys/devices/platform/42120000.i2c/: No such file or directory
IRQ 42120000.i2c(7) guessed as class 0
IRQ 42120000.i2c(7) guessed as class 0
IRQ 40b00000.i2c(9) guessed as class 0
IRQ thr_011(10) guessed as class 0
IRQ 2800000.serial(11) guessed as class 0
IRQ 2000000.i2c(15) guessed as class 0
IRQ 2010000.i2c(16) guessed as class 0
IRQ 2020000.i2c(17) guessed as class 0
IRQ 2030000.i2c(18) guessed as class 0
IRQ mmc0(20) guessed as class 0
IRQ USER1(49) guessed as class 0
IRQ 0-0039(50) guessed as class 0
IRQ USER2(52) guessed as class 0
IRQ 46000000.ethernet(279) guessed as class 5
No directory /sys/devices/platform/46000000.ethernet/: No such file or directory
IRQ 46000000.ethernet(279) guessed as class 0
IRQ 46000000.ethernet(279) guessed as class 0
IRQ 46000000.ethernet(280) guessed as class 5
No directory /sys/devices/platform/46000000.ethernet/: No such file or directory
IRQ 46000000.ethernet(280) guessed as class 0
IRQ 46000000.ethernet(280) guessed as class 0
IRQ 46000000.ethernet(281) guessed as class 5
No directory /sys/devices/platform/46000000.ethernet/: No such file or directory
IRQ 46000000.ethernet(281) guessed as class 0
IRQ 46000000.ethernet(281) guessed as class 0
Adding IRQ 3 to database
Adding IRQ 7 to database
Adding IRQ 9 to database
Adding IRQ 10 to database
Adding IRQ 11 to database
Adding IRQ 15 to database
Adding IRQ 16 to database
Adding IRQ 17 to database
Adding IRQ 18 to database
Adding IRQ 20 to database
Adding IRQ 49 to database
Adding IRQ 50 to database
Adding IRQ 52 to database
Adding IRQ 279 to database
Adding IRQ 280 to database
Adding IRQ 281 to database
NUMA NODE NUMBER: -1
LOCAL CPU MASK: ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,ffffffff,fffffffff




-----------------------------------------------------------------------------
Package 1:  numa_node cpu mask is 0000000c (load 0)
        Cache domain 0:  numa_node is cpu mask is 0000000c  (load 0) 
                CPU number 3  numa_node is (load 0)
                CPU number 2  numa_node is (load 0)
Package 0:  numa_node cpu mask is 00000003 (load 0)
        Cache domain 1:  numa_node is cpu mask is 00000003  (load 0) 
                CPU number 1  numa_node is (load 0)
                CPU number 0  numa_node is (load 0)



-----------------------------------------------------------------------------
Package 1:  numa_node cpu mask is 0000000c (load 0)
        Cache domain 0:  numa_node is cpu mask is 0000000c  (load 0) 
                CPU number 3  numa_node is (load 0)
                CPU number 2  numa_node is (load 0)
Package 0:  numa_node cpu mask is 00000003 (load 0)
        Cache domain 1:  numa_node is cpu mask is 00000003  (load 0) 
                CPU number 1  numa_node is (load 0)
                CPU number 0  numa_node is (load 0)



-----------------------------------------------------------------------------
Package 1:  numa_node cpu mask is 0000000c (load 0)
        Cache domain 0:  numa_node is cpu mask is 0000000c  (load 0) 
                CPU number 3  numa_node is (load 0)
                CPU number 2  numa_node is (load 0)
Package 0:  numa_node cpu mask is 00000003 (load 0)
        Cache domain 1:  numa_node is cpu mask is 00000003  (load 0) 
                CPU number 1  numa_node is (load 10000000)
                CPU number 0  numa_node is (load 10000000)

# for i in $(seq 0 300)
> do
> grep . /proc/irq/$i/smp_affinity /dev/null 2>/dev/null
> done
/proc/irq/1/smp_affinity:f
/proc/irq/2/smp_affinity:f
/proc/irq/3/smp_affinity:f
/proc/irq/4/smp_affinity:f
/proc/irq/5/smp_affinity:f
/proc/irq/7/smp_affinity:f
/proc/irq/9/smp_affinity:f
/proc/irq/10/smp_affinity:f
/proc/irq/11/smp_affinity:f
/proc/irq/15/smp_affinity:f
/proc/irq/16/smp_affinity:f
/proc/irq/17/smp_affinity:f
/proc/irq/18/smp_affinity:f
/proc/irq/20/smp_affinity:f
/proc/irq/49/smp_affinity:f
/proc/irq/50/smp_affinity:f
/proc/irq/52/smp_affinity:f
/proc/irq/279/smp_affinity:f
/proc/irq/280/smp_affinity:f
/proc/irq/281/smp_affinity:f
root@charlie:~# 

root@charlie:~# lstopo-no-graphics
Machine (3984MB)
  Package L#0 + L2 L#0 (512KB)
    L1d L#0 (32KB) + L1i L#0 (32KB) + Core L#0 + PU L#0 (P#0)
    L1d L#1 (32KB) + L1i L#1 (32KB) + Core L#1 + PU L#1 (P#1)
  Package L#1 + L2 L#1 (512KB)
    L1d L#2 (32KB) + L1i L#2 (32KB) + Core L#2 + PU L#2 (P#2)
    L1d L#3 (32KB) + L1i L#3 (32KB) + Core L#3 + PU L#3 (P#3)
root@charlie:~# 
```

**Comment by nhorman** [comment]
Hmm, first curiosity is that there is no message in the CONFIG_NUMA=n case indicating that numa is not available.  theres a call to numa_available in irqbalance that should produce that message if NUMA isn't configured to be on, and I don't see it, leading me to wonder why the numactl library would think numa is available when it isn't.  Can you run irqbalance through gdb and see what numa_available is returning?  It should return a  negative value in the CONFIG_NUMA=n case, and it doesn't appear to be

Second, I don't see any irqs getting selected for rebalancing, which is something that should happen after the first 10 second iteration (which it does in the CONFIG_NUMA=y case).  How long did you let irqbalance run before you checked the interrupts?  It needs to be at least 10 seconds to give irqbalance a chance to compute a workload change and select irqs for rebalancing

**Comment by nsekhar** [comment]
In the CONFIG_NUMA=n case it prints “This machine seems not NUMA capable.”.  Is that not what you are looking for?

The log here is short but I don’t see irqbalance kick in even during long tests with numa disabled in kernel. I can provide a longer log when I get back in office.

**Comment by nhorman** [comment]
Yes, if you can provide the full log in the failure case that would be good.  Also:

1) What kind of system is this?  I can see if I can find a similar machine to reproduce with
2) Can you provide me a tarball of:
  a) /proc/interrupts
  b) /proc/stat
  c) /proc/softirq
  d) /sys

I can use that to try emulate your systems behavior

**Comment by nmenon** [comment]
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/arch/arm64/boot/dts/ti/k3-am65.dtsi is specifically the one @nsekhar is using. cpu topology of interest is there - a GICV3 system without gicv2 backward compatibility.

**Comment by nsekhar** [comment]
@nhorman , attached are the logs you requested.
[irqbalance-logs.tar.gz](https://github.com/Irqbalance/irqbalance/files/2875083/irqbalance-logs.tar.gz). They are all taken in the problematic case (CONFIG_NUMA=n)

**Comment by nhorman** [comment]
Ok, thank you, looking at your logs, you are right, I never see any irqs getting selected for rebalancing, which suggests we need to add some debug code to determine why all the irqs are getting skipped.  I've created this branch:
https://github.com/Irqbalance/irqbalance/tree/arm_debug

With some of that debug code in place, if you could please clone, build and test it in the failling environment and report the full log back here, I would apprecaite it.

**Comment by nsekhar** [comment]
Hi @nhorman, here are the logs from your branch.
[irqbalance-arm_debug.txt](https://github.com/Irqbalance/irqbalance/files/2878849/irqbalance-arm_debug.txt)

**Comment by nsekhar** [comment]
@nhorman , In my earlier log, I did not have libnuma-dev installed, so irqbalance detected NUMA not supported whether CONFIG_NUMA was enabled or not. Here is a log after I compiled again with libnuma-dev installed, but CONFIG_NUMA disabled in kernel. I do not think the log is any different, but this represents the original issue better.
[irqbalance-arm_debug-v2.txt](https://github.com/Irqbalance/irqbalance/files/2879362/irqbalance-arm_debug-v2.txt)

**Comment by nhorman** [comment]
Interesting, based on what you've sent it, it appears as though we have a problem in which irqs never get assigned to a numa node when numa isn't available, which means they never filter down to a more precise package/cache domain/cpu (in the event numa isn't available we are supposed to have a 'dummy node' that represents all cpus and irqs). I'm not sure why its happening yet, but on the upside, I can reproduce the problem if I force numa off on my x86 system.  Let me debug it here, and I'll report back as soon as I have a fix

**Comment by nsekhar** [comment]
@nhorman nice to hear you are able to see the issue too. Looking forward to a fix!

**Comment by nhorman** [comment]
Ok, I've pushed a patch that adds the unspecified dummy numa node to the internal device tree when numa isn't available, so that irqs properly get assigned down through the device tree to the proper level.  Its fixed it on my reproducer.  If you could please, give it a spin on your arm system in both the numa available and numa not available cases and confirm that it works for you.  Thanks!

**Comment by nsekhar** [comment]
Thanks for the fix! I am not in office next 2 days, so will be able to test this on Monday.

**Comment by nhorman** [comment]
That's fine, I'm actually out until Monday myself.

Thanks!
Neil

**Comment by AndCycle** [comment]
I have a non direct related report for this, 
as I accidentally got my kernel on x86 have `NUMA=n`, 
confirmed `arm_debug` branch fixed the behavior on it.

**Comment by nhorman** [comment]
thank you @AndCycle .  If @nsekhar confirms, I'll merge this today

**Comment by nsekhar** [comment]
@nhorman , I was able to get to testing this today. The patch indeed works for me too. Logs here: https://pastebin.ubuntu.com/p/mxHdXmjdVm/

**Comment by nhorman** [comment]
awesome, thanks!  I'll integrate this today then.

**Comment by nhorman** [comment]
merged as commit 9f7b1a93150c6bb1f54a7bf5ce36b344f10cdf4a
